src/mame/drivers/tsconf.cpp

```diff
@@ -0,0 +1,293 @@
+ * - Sprites
+ * - Sound
+ * - ZX-Mode locks
+ * - CPU frequency
+ * - Timings
+ * - Ram cache
+ * - VDos
+ * - Many more...
+ * 
+ ****************************************************************************/
+
+#include "includes/tsconf.h"
+
+TILE_GET_INFO_MEMBER(tsconf_state::get_tile_info_txt)
+{
+	u8 *messram = m_ram->pointer();
+	u8 *m_row_location = messram + PAGE(V_PAGE) + (tile_index / tilemap.cols() * 256);
+	u8 col = tile_index % tilemap.cols();
+	u8 symbol = m_row_location[col];
+	tileinfo.set(1, symbol, 0, 0);
+}
+
+template <u8 Layer>
+TILE_GET_INFO_MEMBER(tsconf_state::get_tile_info_16c)
+{
+	u8 col_offset = (tile_index % tilemap.cols()) << 1;
+	u16 row_offset = (((tile_index / tilemap.cols()) << 1) + Layer) * 64 * 2;
+
+	u8 *messram = m_ram->pointer();
+	u8 *tile_info_addr = messram + PAGE(T_MAP_PAGE) + row_offset + col_offset;
+	u8 hi = tile_info_addr[1];
+
+	u16 tile = ((u16(hi) & 0x0f) << 8) | tile_info_addr[0];
+	tile = tile / tilemap.cols() * 64 * 8 + (tile % tilemap.cols());
+	u8 pal = ((BIT(PAL_SEL, 4 + Layer * 2, 2) << 2) | BIT(hi, 4, 2)) << 4;
+	if (BIT(hi, 6, 2))
+	{
+		LOGWARN("!!! - FLIP Case\n");
+	}
+	tileinfo.set(2 + Layer, tile, u16(pal), 0);
+}
+
+void tsconf_state::tsconf_mem(address_map &map)
+{
+	map(0x0000, 0x3fff).bankr("bank1").w(FUNC(tsconf_state::tsconf_bank_w<0>));
+	map(0x4000, 0x7fff).bankr("bank2").w(FUNC(tsconf_state::tsconf_bank_w<1>));
+	map(0x8000, 0xbfff).bankr("bank3").w(FUNC(tsconf_state::tsconf_bank_w<2>));
+	map(0xc000, 0xffff).bankr("bank4").w(FUNC(tsconf_state::tsconf_bank_w<3>));
+}
+
+void tsconf_state::tsconf_io(address_map &map)
+{
+	map.unmap_value_high();
+	map(0x0000, 0x0000).w(FUNC(tsconf_state::tsconf_port_7ffd_w)).mirror(0x7ffd);
+	map(0x001f, 0x001f).rw(m_beta, FUNC(beta_disk_device::status_r), FUNC(beta_disk_device::command_w)).mirror(0xff00);
+	map(0x003f, 0x003f).rw(m_beta, FUNC(beta_disk_device::track_r), FUNC(beta_disk_device::track_w)).mirror(0xff00);
+	map(0x0057, 0x0057).rw(FUNC(tsconf_state::tsconf_port_57_zctr_r), FUNC(tsconf_state::tsconf_port_57_zctr_w)).mirror(0xff00); // spi config
+	map(0x0077, 0x0077).rw(FUNC(tsconf_state::tsconf_port_77_zctr_r), FUNC(tsconf_state::tsconf_port_77_zctr_w)).mirror(0xff00); // spi data
+	map(0x005f, 0x005f).rw(m_beta, FUNC(beta_disk_device::sector_r), FUNC(beta_disk_device::sector_w)).mirror(0xff00);
+	map(0x007f, 0x007f).rw(m_beta, FUNC(beta_disk_device::data_r), FUNC(beta_disk_device::data_w)).mirror(0xff00);
+	map(0x00fe, 0x00fe).rw(FUNC(tsconf_state::spectrum_port_fe_r), FUNC(tsconf_state::spectrum_port_fe_w)).select(0xff00);
+	map(0x00ff, 0x00ff).rw(m_beta, FUNC(beta_disk_device::state_r), FUNC(beta_disk_device::param_w)).mirror(0xff00);
+	map(0x00af, 0x00af).rw(FUNC(tsconf_state::tsconf_port_xxaf_r), FUNC(tsconf_state::tsconf_port_xxaf_w)).select(0xff00);
+	map(0x8ff7, 0x8ff7).w(FUNC(tsconf_state::tsconf_port_f7_cmos_w)).select(0x7000); // 3:bff7 5:dff7 6:eff7
+	map(0xbff7, 0xbff7).r(FUNC(tsconf_state::tsconf_port_f7_cmos_r));
```

Comment:
I’d prefer if the `select` and `mirror` stuff went before the handlers, so the addressing is all visible at the start of the lines.